### PR TITLE
 Add Support for 4 port VICI Valve

### PR DESCRIPTION
**Pull request by jozamudi**
## Description

FB_ViciValve: Bundles FB Components needed to communicate with valve though RS-232 interface for Vici Valves.
FB_4PViciValve: Function Block for 4 Port Vici Valve
FB_4PViciValveSampleNames: Contains PVs for Port Labels.

## Motivation and Context

https://jira.slac.stanford.edu/browse/ECS-2140

## How Has This Been Tested?

A 4 port vici valve was setup in ICL and changed port positions though EPICS PVs.

## Screenshots (if appropriate):
image

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive comments
- [x] Test suite passes locally
- [x] Libraries are set to fixed versions and not ``Always Newest``
- [x] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

### CODE REVIEW

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+     stSampleName : ST_4PViciValveSampleNames;
+
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Re-using 12 Port Code, Need to map 1-4 to vlaues that 12 port Commands
+//map 1-4 to 9 - 12;
+
+iNewPos := LIMIT(1, iVici1ReqPos, 4);
+iMaped := 12 + ((iNewPos - 1)*(9 - 12) / (4 - 1));
```
Comment (ZLLentz) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU]: Can this be rewritten more straightforwardly for maintainability purposes? It looks like the effect here is to map 1 -> 12, 2 -> 11, 3 -> 10, and 4 -> 9?

So maybe just

```
iMaped := 13 - iNewPos
```

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        pv: SampleNames
+        io: io
+     '}
+     stSampleName : ST_4PViciValveSampleNames;
+
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Re-using 12 Port Code, Need to map 1-4 to vlaues that 12 port Commands
+//map 1-4 to 9 - 12;
```
Comment (ZLLentz) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU]: What's the story behind this mapping? Where's the documentation you read to decide on this mapping?

Comment (jozamudi) [reply lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU]: There Is a code block in the [FB_VICIDriver ](https://github.com/pcdshub/lcls-twincat-sample-delivery/blob/2eab9f0c6295b374a5f2c6be21e9916c6d16f390/lcls-twincat-sample-delivery/Library/POUs/VICI/FB_ViciDriver.TcPOU#L94).  

There is a conversion that comments (* Mirror the "clock" value across the vertical axis of symmetry *). This means that the port requested does not equal to the command sent to the valve. Requesting port 1 sends command  for the port to go to the 12th port.  The mapping used in the 4port FB to match the requested port number with the appropriate command for the 4 port valve.

**Comment by KaushikMalapati** [comment]
I think ST_4PViciValveSampleNames could be replaced by defining stSampleNames as 
```
{attribute 'pytmc' := '
        pv: SampleNames
        io: io
     '}
     stSampleNames : ARRAY [1..4] of STRING;
``` 
According to https://pcdshub.github.io/pytmc/v2.16.0/pragma_usage.html#arrays, this will generate that same pvs that you manually defined in ST_4PViciValveSampleNames.

**Comment by KaushikMalapati** [comment]
Also, this could easily generalized to work with n-ports by having an in var that you use as the upper bound of the limit on line 41. Alternatively, you could put the upper bound that corresponds to the number of ports in FB_ViciDriver in this area 
```
    5000: (* Set position*)
        (* Limit the requested position to the valid 12 possible positions *)
        IF i_stControl.iReqPos > 12 OR i_stControl.iReqPos < 1 THEN
            q_sResult := CONCAT('in step 5000 port range exceeded: ',FBViciTransaction.q_sResponseData);
            i_stControl.iReqPos := LIMIT(1, i_stControl.iReqPos,12);
``` 
Replacing the hard coded 12 with an in var. This way we also get the message added to q_sResult. Then this function block could just pass the number of ports to FB_ViciDriver, and this POU would be for general continuous use of a valve. Anywhere we want to have the driver loop like here in FB_SelectorM3
```
afbViciDriver[viciIndex](
    i_xExecute:= TRUE,
    i_tTimeOut:= t#10s,
    i_stControl:= stSelector.astViciVlvCtrl[viciIndex],
    iq_stSerialRXBuffer:= stViciSerial.RxBuffer,
    iq_stSerialTXBuffer:= stViciSerial.TxBuffer,
    q_stStatus=>stSelector.astViciVlvStatus[viciIndex]);

IF 	afbViciDriver[viciIndex].q_xDone OR
    afbViciDriver[viciIndex].q_xError OR
    afbViciDriver[viciIndex].q_xTimeout THEN
    (* reset function for next time *)
    afbViciDriver[viciIndex](i_xExecute:=FALSE,
        iq_stSerialRXBuffer:= stViciSerial.RxBuffer,
        iq_stSerialTXBuffer:= stViciSerial.TxBuffer);
``` 
We would just call your new POU

**Comment by jozamudi** [comment]
I tried to create the string array but the IOC fails to build. 

ERROR:pytmc.bin.template:Failed to create EPICS records
Traceback (most recent call last):
  File "c:\repos\pytmc\pytmc\bin\template.py", line 374, in get_plc_record_packages
    packages, exceptions = db_process(
  File "c:\repos\pytmc\pytmc\bin\db.py", line 154, in process
    record_names = [
  File "c:\repos\pytmc\pytmc\bin\db.py", line 158, in <listcomp>
    for single_record in record_package.records
  File "c:\repos\pytmc\pytmc\record.py", line 619, in records
    records = [self.generate_input_record()]
  File "c:\repos\pytmc\pytmc\record.py", line 955, in generate_input_record
    record = super().generate_input_record()
  File "c:\repos\pytmc\pytmc\record.py", line 511, in generate_input_record
    record.fields["DTYP"] = self.dtyp
  File "c:\repos\pytmc\pytmc\record.py", line 952, in dtyp
    return data_types[self.chain.data_type.name]
KeyError: 'STRING'

**Comment by ZLLentz** [comment]
It looks like pytmc there is trying to create a literal array of strings (which has dubious support in EPICS, and probably should fail).
Remembering a bit from when this was implemented, the reason why arrays of structs map to separate, numbered PVs is because there isn't a way to pack them into EPICS arrays, which is the default thing pytmc tries to do.

IMO pytmc should treat arrays of strings as a special case as Kaushik expected earlier rather than erroring out entirely.
Arrays of e.g. integers should probably map to integer waveforms still.
Maybe this needs to have some customization options.

**Comment by jozamudi** [comment]
I did consider making the changes Kaushik pointed out, but I was also reluctant to edit the FB_ViciDriver directly since this Function block is being used in other Projects.  Writing a wrapper FB seemed like a good solution.

**Comment by ZLLentz** [comment]
I think it's OK to implement things in easy, noninvasive ways and deploy them as part of PLC projects. I think it's also a good idea to make library updates that are forward-thinking when we bring these changes up into shared, re-usable library code.

**Comment by KaushikMalapati** [comment]
I think I missed this earlier, but the thirteen on line 86 in ViciDriver should also be changed to i_stControl.iPortCount
```
q_stStatus.iCurrPos := 13 - STRING_TO_INT(sWorking); (* pass that INT to the q_stStatus.iCurrPos, mirroring positions *)
```

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        pv: ReqPos
+        io: io
+    '}
+    iVici1ReqPos : INT;
+
+    {attribute 'pytmc' := '
+        pv: SampleNames
+        io: io
+     '}
+     stSampleName : ST_4PViciValveSampleNames;
```
Comment (KaushikMalapati) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PViciValve.TcPOU]:
```suggestion
     stSampleNames : ST_4PViciValveSampleNames;
```

**Comment by KaushikMalapati** [comment]
Everything else looks good to me

**Comment by jozamudi** [comment]
Would it make sense to add a generic FB_ViciValve Function block and let the number of port and Sample name PVs  defined in the PLC application?

**Comment by KaushikMalapati** [comment]
I think it would have been cleaner to make iportCount a VAR_IN  FB_4PViciValve.TcPOU, rename the file to something like FB_NPortViciValve and not have a specific POU for the four valve. Then in the plc project, you just instantiate the POU with iportCount equal to however many valves you have. If we are using 4 port valves often , then we can just extend FB_NPortViciValve with iportCount equal to four

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU [hunk 4]
```diff
@@ -0,0 +1,30 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4024.12">
+  <POU Name="FB_4PortViciVavle" Id="{35f5328b-2692-4fc5-a1f1-07aefd5e56d3}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK INTERNAL FB_4PortViciVavle
+VAR_INPUT
+    iAdress : INT;
```
Comment (KaushikMalapati) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU]:
```suggestion
    iAddress : INT;
```

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+END_VAR
+VAR_OUTPUT
+END_VAR
+VAR
+    fbNporValve : FB_ViciValve;
```
Comment (KaushikMalapati) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU]:
```suggestion
    fbNportValve : FB_ViciValve;
```

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    iReqPos : INT;
+    {attribute 'pytmc' := '
+        pv: SampleNames
+        io: io
+     '}
+    stSampleNames : ST_4PViciValveSampleNames;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[fbNporValve(iAddress := iAdress, iNumOfPorts := iNumPorts, iReqPos := iReqPos);]]></ST>
```
Comment (KaushikMalapati) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU]:
```suggestion
      <ST><![CDATA[fbNportValve(iAddress := iAdress, iNumOfPorts := iNumPorts, iReqPos := iReqPos);]]></ST>
```

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU [hunk 7]
```diff
@@ -0,0 +1,30 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4024.12">
+  <POU Name="FB_4PortViciVavle" Id="{35f5328b-2692-4fc5-a1f1-07aefd5e56d3}" SpecialFunc="None">
```
Comment (KaushikMalapati) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU]:
```suggestion
  <POU Name="FB_4PortViciValve" Id="{35f5328b-2692-4fc5-a1f1-07aefd5e56d3}" SpecialFunc="None">
```

File: lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    <Declaration><![CDATA[FUNCTION_BLOCK INTERNAL FB_4PortViciVavle
```
Comment (KaushikMalapati) [inline lcls-twincat-sample-delivery/Library/POUs/VICI/FB_4PortViciVavle.TcPOU]:
```suggestion
    <Declaration><![CDATA[FUNCTION_BLOCK INTERNAL FB_4PortViciValve
```

**Comment by KaushikMalapati** [comment]
The file FB_4PortViciVavle.TcPou should be renamed to FB_4PortViciValve.TcPou. 

Also, I think the documentation error is because Library.tsproj still has a reference to FB_4PViciValve.TcPou instead of FB_4PortViciVavle.TcPOU. Maybe you forgot to save this after updating those files, or you did not add and commit it? It also is missing a reference to FB_ViciValve.TcPOU.

Everything else looks solid to me, assuming you have already tested this in ICL.

**Review by KaushikMalapati — APPROVED** [review]

**Pull request merged**